vcd/resource_vcd_vapp_vm.go

```diff
@@ -168,38 +282,97 @@ func resourceVcdVAppVmCreate(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("error finding vApp: %#v", err)
 	}
 
-	var network *types.OrgVDCNetwork
+	// Networking part
+
+	// Step 1 - gather vars
+	legacyNetworkName := d.Get("network_name").(string)
+	legacyVappNetworkName := d.Get("vapp_network_name").(string)
+
+	allNetworks := d.Get("networks").([]interface{})
+
+	var vAppNets []*types.OrgVDCNetwork
+	var vdcNets []*types.OrgVDCNetwork
 
-	if d.Get("network_name").(string) != "" {
-		network, err = addVdcNetwork(d, vdc, vapp, vcdClient)
+	// Step 2 - process legacy "network_name" and "vapp_network_name"
+	// TODO 3.0 remove when "network_name" and "vapp_network_name" is deprecated
+	if legacyNetworkName != "" {
+		legacyNetwork, err := addVdcNetwork(d.Get("network_name").(string), vdc, vapp, vcdClient)
 		if err != nil {
 			return err
 		}
+		vdcNets = append(vdcNets, legacyNetwork)
 	}
-
-	vappNetworkName := d.Get("vapp_network_name").(string)
-	if vappNetworkName != "" {
-		isVappNetwork, err := isItVappNetwork(vappNetworkName, vapp)
+	if legacyVappNetworkName != "" {
+		isVappNetwork, err := isItVappNetwork(legacyVappNetworkName, vapp)
 		if err != nil {
 			return err
 		}
 		if !isVappNetwork {
-			return fmt.Errorf("vapp_network_name: %s is not found", vappNetworkName)
+			return fmt.Errorf("vapp_network_name: %s is not found", legacyVappNetworkName)
 		}
 	}
+	// TODO 3.0 EO remove when "network_name" and "vapp_network_name" is deprecated
 
-	err = retryCall(vcdClient.MaxRetryTimeout, func() *resource.RetryError {
-		log.Printf("[TRACE] Creating VM: %s", d.Get("name").(string))
-		var networks []*types.OrgVDCNetwork
-		if network != nil {
-			networks = append(networks, network)
+	// Step 3 - process new networks of all types
+	for _, network := range allNetworks {
+		net := network.(map[string]interface{})
+		netType := net["network_type"].(string)
+		netName := net["network_name"].(string)
+
+		switch netType {
+		case "vapp":
+			isVappNetwork, err := isItVappNetwork(netName, vapp)
+			if err != nil {
+				return fmt.Errorf("unable to find vApp network %s: %s", netName, err)
+			}
+			if !isVappNetwork {
+				return fmt.Errorf("vApp network : %s is not found", netName)
+			}
+			vAppNets = append(vAppNets, &types.OrgVDCNetwork{Name: netName})
+		case "org":
+			vdcNet, err := vdc.FindVDCNetwork(netName)
+			if err != nil {
+				return fmt.Errorf("error finding org vdc network: %#v", err)
+			}
+			vdcNets = append(vdcNets, vdcNet.OrgVDCNetwork)
+		case "none":
+			// Special reference type 'none' network  does not exist anywhere, but allows to add unattached NIC.
+			// Store empty object with hardcoded type.
+			vdcNets = append(vdcNets, &types.OrgVDCNetwork{Name: types.NoneNetwork})
 		}
-		task, err := vapp.AddVM(networks, vappNetworkName, vappTemplate, d.Get("name").(string), acceptEulas)
 
+	}
+
+	// Step 4 - ensure all specified vdc networks are attached to vApp
```

Comment:
If you want comment with steps is strong indication that function should spitted to meaningful parts with self explanatory functions